### PR TITLE
store/tikv: more friendly error log for WriteConflict

### PR DESCRIPTION
**Pull request by tiancaiamao**
Instead of 
```
key: [116, 128, 0, 0, 0, 0, 0, 1, 255, 155, 95, 105, 128, 0, 0, 0, 0, 255, 0, 0, 1, 1, 82, 87, 48, 49, 255, 0, 0, 0, 0, 251, 1, 55, 54, 255, 56, 50, 50, 49, 49, 48, 255, 57, 255, 0, 0, 0, 0, 0, 0, 0, 248, 255, 1, 0, 0, 0, 0, 0, 0, 0, 255, 0, 247, 0, 0, 0, 0, 0, 0, 249], 
primary: [116,128, 0, 0, 0, 0, 0, 1, 155, 95, 105, 128, 0, 0, 0, 0, 0, 0, 1, 1, 82, 87, 48, 49, 0, 0, 0, 0, 251, 1, 55, 54, 56, 50, 50, 49, 49, 48, 255, 57, 0, 0, 0, 0, 0, 0, 0, 248, 1, 0, 0, 0, 0, 0, 0, 0, 0, 247] }))
```

We can try to decode the key information to make it more human readable.

```
primary={tableID=411, indexID=1, indexValues={6-[82 87 48 49], 6-[55 54 56 50 50 49 49 48 57], 6-[], }}
```

Depends on https://github.com/pingcap/kvproto/pull/252

@coocood

### CODE REVIEW

**Comment by shenli** [comment]
We need to update kvproto in the deps first?

**Comment by tiancaiamao** [comment]
Yes.

**Comment by shenli** [comment]
So we need to wait another PR?

**Comment by tiancaiamao** [comment]
/run-all-tests
PTAL @coocood @shenli

File: store/tikv/snapshot.go [hunk 1]
```diff
@@ -296,6 +299,10 @@ func extractLockFromKeyErr(keyErr *pb.KeyError) (*Lock, error) {
 	if locked := keyErr.GetLocked(); locked != nil {
 		return NewLock(locked), nil
 	}
+	if keyErr.Conflict != nil {
+		err := errors.Errorf(conflictToString(keyErr.Conflict))
```
Comment (zimulala) [inline store/tikv/snapshot.go]: s/Errorf/New ?

File: store/tikv/snapshot_test.go [hunk 2]
```diff
@@ -141,3 +142,15 @@ func makeKeys(rowNum int, prefix string) []kv.Key {
...
+		ConflictTs: 399402937719455772,
+		Key:        []byte{116, 128, 0, 0, 0, 0, 0, 1, 255, 155, 95, 105, 128, 0, 0, 0, 0, 255, 0, 0, 1, 1, 82, 87, 48, 49, 255, 0, 0, 0, 0, 251, 1, 55, 54, 255, 56, 50, 50, 49, 49, 48, 255, 57, 255, 0, 0, 0, 0, 0, 0, 0, 248, 255, 1, 0, 0, 0, 0, 0, 0, 0, 255, 0, 247, 0, 0, 0, 0, 0, 0, 249},
+		Primary:    []byte{116, 128, 0, 0, 0, 0, 0, 1, 155, 95, 105, 128, 0, 0, 0, 0, 0, 0, 1, 1, 82, 87, 48, 49, 0, 0, 0, 0, 251, 1, 55, 54, 56, 50, 50, 49, 49, 48, 255, 57, 0, 0, 0, 0, 0, 0, 0, 248, 1, 0, 0, 0, 0, 0, 0, 0, 0, 247},
+	}
+
+	str := `WriteConflict: startTS=399402937522847774, conflictTS=399402937719455772, key=[]byte{0x74, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xff, 0x9b, 0x5f, 0x69, 0x80, 0x0, 0x0, 0x0, 0x0, 0xff, 0x0, 0x0, 0x1, 0x1, 0x52, 0x57, 0x30, 0x31, 0xff, 0x0, 0x0, 0x0, 0x0, 0xfb, 0x1, 0x37, 0x36, 0xff, 0x38, 0x32, 0x32, 0x31, 0x31, 0x30, 0xff, 0x39, 0xff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf8, 0xff, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0x0, 0xf7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf9} primary={tableID=411, indexID=1, indexValues={6-[82 87 48 49], 6-[55 54 56 50 50 49 49 48 57], 6-[], }}`
+	c.Assert(conflictToString(conflict), Equals, str)
```
Comment (zimulala) [inline store/tikv/snapshot_test.go]: Do we need to add a test for `RecordKey`?

File: store/tikv/snapshot.go [hunk 3]
```diff
@@ -308,3 +315,32 @@ func extractLockFromKeyErr(keyErr *pb.KeyError) (*Lock, error) {
...
+	prettyWriteKey(&buf, conflict.Primary)
+	return buf.String()
+}
+
+func prettyWriteKey(buf *bytes.Buffer, key []byte) {
+	tableID, indexID, indexValues, err := tablecodec.DecodeIndexKey(key)
+	if err == nil {
+		fmt.Fprintf(buf, "{tableID=%d, indexID=%d, indexValues={", tableID, indexID)
+		for _, v := range indexValues {
+			fmt.Fprintf(buf, "%s, ", v)
```
Comment (coocood) [inline store/tikv/snapshot.go]: How about using `ToString` to print value?

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
What's happening about the golint

**Comment by lysu** [comment]
https://github.com/kisielk/errcheck/commit/e5700e7d6efa0cc80bd13cbc2607c728fa553d01, It seems that `fmt.xxx` expect print to stdout need checkerr now - -

**Comment by coocood** [comment]
@tiancaiamao 
It's not golint, it's errcheck, `Fprintf` returns error and we don't check it.

**Comment by coocood** [comment]
LGTM

**Pull request merged**